### PR TITLE
esp32: control the python heap size via NVS variables

### PR DESCRIPTION
**Pull request by tve**
This PR builds on top of #6780 and #6782.

This commit allows the user to limit the micropython heap size by setting NVS variables. This supports:
- ensuring that esp-idf has enough memory for multiple TLS connections, TLS plus BLE, etc.
- leaving some SPIRAM memory for native modules, such as camera framebuffers
- reducing the SPIRAM memory to limit GC sweep times

See the docs changes for more rationale and details.
A test script is included, but not usable with run-tests due to the resets required.

### CODE REVIEW

**Comment by pumelo** [comment]
Cool, like it.
What is the maximum heap size you can allocate for mp with this method? I briefly recall, that I have tried something similar but ended up with a much smaller max heap size for mp than when just using all SPIRAM. What is your expirience?

**Comment by tve** [comment]
> What is the maximum heap size you can allocate for mp with this method?

The max size is what you get today and continues to be the same if you do not set any NVS variable. This PR only allows you to _reduce_ the heap from the default. It does not allow you to grab more memory. (MP already grabs the max it can...)

**Comment by pumelo** [comment]
If you follow this code path starting from here:
https://github.com/micropython/micropython/blob/1f800cac3c4f56418486f24c3824418a890f759f/ports/esp32/main.c#L82
Here malloc is not used but the whole spiram assigned to mp. using malloc results in a much smaller amount of heap assigned to mp (at least in my expirience). That was the reason asking.

**Comment by tve** [comment]
> If you follow this code path starting from here:
> https://github.com/micropython/micropython/blob/1f800cac3c4f56418486f24c3824418a890f759f/ports/esp32/main.c#L82
> 
> 
> Here malloc is not used but the whole spiram assigned to mp. using malloc results in a much smaller amount of heap assigned to mp (at least in my expirience). That was the reason asking.

The trick is `CONFIG_SPIRAM_USE_CAPS_ALLOC=y`. Before, the SPIRAM was hidden from esp-idf's malloc. With the changes it no longer is.

**Comment by dpgeorge** [comment]
Can you please rebase this on latest master?  All prerequisites are now merged.

**Comment by tve** [comment]
Rebased against current master and tested against esp-idf v4.2 using GENERIC and TINYPICO board variants.

File: ports/esp32/main.c [hunk 1]
```diff
@@ -75,30 +75,43 @@ void mp_task(void *pvParameter) {
...
+            if (max_mp_heap_size > MIN_HEAP_SIZE && mp_task_heap_size > max_mp_heap_size) {
+                // we're about to create too large a heap, be more modest
+                mp_task_heap_size = max_mp_heap_size;
+            }
+        }
+        nvs_close(mp_nvs);
+    }
+    void *mp_task_heap = heap_caps_malloc(mp_task_heap_size, MALLOC_CAP_8BIT);
+    if (avail_heap_size != mp_task_heap_size) {
+        printf("Heap limited: avail=%d, actual=%d, left to idf=%d\n",
```
Comment (dpgeorge) [inline ports/esp32/main.c:112]: IMO it should not print anything here.  The user can programatically inspect the heap sizes and print their own message if needed.

File: ports/esp32/modesp32.c [hunk 2]
```diff
@@ -190,6 +190,7 @@ STATIC const mp_rom_map_elem_t esp32_module_globals_table[] = {
     { MP_ROM_QSTR(MP_QSTR_Partition), MP_ROM_PTR(&esp32_partition_type) },
     { MP_ROM_QSTR(MP_QSTR_RMT), MP_ROM_PTR(&esp32_rmt_type) },
     { MP_ROM_QSTR(MP_QSTR_ULP), MP_ROM_PTR(&esp32_ulp_type) },
+    { MP_ROM_QSTR(MP_QSTR_NVS), MP_ROM_PTR(&esp32_nvs_type) },
```
Comment (dpgeorge) [inline ports/esp32/modesp32.c:193]: not needed

File: ports/esp32/modesp32.h [hunk 3]
```diff
@@ -30,5 +30,6 @@ extern const mp_obj_type_t esp32_nvs_type;
 extern const mp_obj_type_t esp32_partition_type;
 extern const mp_obj_type_t esp32_rmt_type;
 extern const mp_obj_type_t esp32_ulp_type;
+extern const mp_obj_type_t esp32_nvs_type;
```
Comment (dpgeorge) [inline ports/esp32/modesp32.h:33]: not needed

**Comment by dpgeorge** [comment]
This is a useful feature.  But I'm wondering if it could be done using normal Python function calls rather than NVS settings, which would then be generalisable to other ports that may want this feature.

The reason stm32 had boot.py and main.py split from the beginning was to configure things like this (USB settings, enable certain hardware features) which affect the system in a significant way.  For example we could add something like:
```python
micropython.resize_heap(lower_bound, upper_bound)
```
which restarts the whole system with a resized heap if its size is not already within the given bounds.  (And maybe eventually this function could be used it increase heap size on the fly, eg on unix where it can ask the OS for more memory.)

In the end, both NVS settings and Python function calls scale well to many configuration settings.  It's just that Python function calls are more Pythonic and could be used on other ports.

**Comment by jimmo** [comment]
We'd like to solve this using an approach similar to described above with boot.py. In particular we now support split heap, so the heap can grow at runtime.
Ref https://github.com/micropython/micropython/issues/8940 for more background on IDF heap issues.

**Comment by tve** [comment]
For reference: split heap PR https://github.com/micropython/micropython/pull/8526

**Pull request closed without merging**